# Homework 1

In [5]:
## Set up the enviroment
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Question 1
What's the version of NumPy that you installed?

In [6]:
np.__version__

'1.22.4'

## Question 2
How many records are in the dataset?

In [8]:
## Let't get the data first
df=pd.read_csv('data.csv')
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [9]:
df.shape

(11914, 16)

This dataset has 11914 records and 16 features

## Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [11]:
df.Make.value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
McLaren             5
Alfa Romeo          5
Genesis   

The most frequent car manufacturers are Chevrolet, Ford, Volkswagen

## Question 4
What's the number of unique Audi car models in the dataset?

In [16]:
len(df[df.Make=='Audi']['Model'].unique())

34

there are 34 unique Audi car models in the dataset

## Question 5
How many columns in the dataset have missing values?

In [23]:
Mss_data=df.isna().sum()
Mss_data[Mss_data>0]

Engine Fuel Type       3
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3742
dtype: int64

5 columns has missing values

## Question 6
1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

In [24]:
#1
df['Engine Cylinders'].median()

6.0

In [27]:
#2
df['Engine Cylinders'].mode()

# 4.0 is the most frequent value

0    4.0
dtype: float64

In [28]:
#3 y 4
df['Engine Cylinders']=df['Engine Cylinders'].fillna(4.0)
df['Engine Cylinders'].median()

6.0

the median remains the same

## Question 7

1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the value of the first element of w?


In [29]:
#1
df_lotus=df[df.Make=='Lotus']

In [31]:
# 2, 3
df_lotus=df_lotus[['Engine HP','Engine Cylinders']]
df_lotus.drop_duplicates(inplace=True)
df_lotus

<ipython-input-31-d586ccc46331>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_lotus.drop_duplicates(inplace=True)


Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [40]:
#4, 5 y 6
X=df_lotus.values
XTX=np.matmul(X.T,X)
XTX=np.linalg.inv(XTX)

In [41]:
XTX

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [42]:
# 7
y=np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [43]:
#8 y 9
w=np.matmul(XTX,X.T)
w=np.matmul(w,y)

In [44]:
w

array([  4.59494481, -63.56432501])

The first element of w is 4.59494481